nvflare/client/model_registry.py

```diff
@@ -73,75 +36,94 @@ class ModelRegistry:
 
     """
 
-    def __init__(self, config: ClientConfig, rank, model_exchanger: Optional[ModelExchanger] = None):
-        self.model_exchanger = model_exchanger
+    def __init__(self, config: ClientConfig, rank, data_exchanger: Optional[DataExchanger] = None):
+        self.data_exchanger = data_exchanger
         self.config = config
 
-        self.cached_model: Optional[FLModel] = None
+        self.received_task: Optional[ExchangeTask] = None
+        self.task_name: str = ""
         self.cache_loaded = False
         self.metrics = None
         self.sys_info = {}
         for k, v in self.config.config.items():
             if k in SYS_ATTRS:
                 self.sys_info[k] = v
-        self.output_meta = {}
         self.rank = rank
 
-    def receive(self):
-        data_file_path = get_data_file_name(self.config.config)
-        if not self.model_exchanger:
-            current_round = read_data(data_file_path)
-            self._set_model(FLModel(current_round=current_round))
-        else:
-            received_model = self.model_exchanger.receive_model()
-            self._set_model(received_model)
-            # write out only the current_round for other ranks
-            save_data_atomic(received_model.current_round, data_file_path)
-
-    def _set_model(self, model: FLModel):
-        self.cached_model = model
+    def receive(self, timeout: Optional[float] = None):
+        if not self.data_exchanger:
+            return None
+
+        task_name, task = self.data_exchanger.receive_data(timeout)
+
+        if task is None:
+            raise RuntimeError("no received task")
+
+        if task.data is None:
+            raise RuntimeError("no received model")
+
+        self.received_task = task
+        self.task_name = task_name
         self.cache_loaded = True
 
-    def get_model(self):
-        if not self.cache_loaded:
-            self.receive()
-        return copy.deepcopy(self.cached_model)
+    def get_task(self, timeout: Optional[float] = None) -> Optional[ExchangeTask]:
```

Comment:
How do you ensure that the received data is type ExchangeTask?

Reply:
added checks